src/signalrclient/cancellation_token_source.h

```diff
@@ -7,57 +7,79 @@
 #include <assert.h>
 #include <condition_variable>
 #include <mutex>
+#include <vector>
+#include "signalrclient/cancellation_token.h"
 
 namespace signalr
 {
     class canceled_exception : public std::exception
     {
     };
 
-    class cancellation_token
+    class cancellation_token_source
     {
     public:
-        static const unsigned int timeout_infinite = 0xFFFFFFFF;
+        static const unsigned int timeout_infinite;
 
-        cancellation_token() noexcept
+        cancellation_token_source() noexcept
             : m_signaled(false)
         {
         }
 
-        cancellation_token(const cancellation_token&) = delete;
-        cancellation_token& operator=(const cancellation_token&) = delete;
+        cancellation_token_source(const cancellation_token_source&) = delete;
+        cancellation_token_source& operator=(const cancellation_token_source&) = delete;
 
-        ~cancellation_token()
+        ~cancellation_token_source()
         {
-            if (m_callback)
+            if (!m_callbacks.empty())
             {
-                m_callback();
+                for (auto& func : m_callbacks)
+                {
+                    try
+                    {
+                        func();
+                    }
+                    catch (...)
+                    {
+                        assert(false);
+                    }
+                }
             }
         }
 
         void cancel()
         {
-            std::function<void()> callback;
+            std::vector<std::function<void()>> callbacks;
             {
                 std::lock_guard<std::mutex> lock(m_lock);
                 m_signaled = true;
                 m_condition.notify_all();
-                callback = m_callback;
-                m_callback = nullptr;
+                callbacks = std::move(m_callbacks);
+                m_callbacks = std::vector<std::function<void()>>();
             } // unlock
 
-            if (callback)
+            if (!callbacks.empty())
             {
-                callback();
+                for (auto& func : callbacks)
+                {
+                    try
+                    {
+                        func();
+                    }
+                    catch (...)
+                    {
+                        assert(false);
```

Comment:
Let's follow the pattern from C#'s CTS and rethrow these later at the end of `cancel()` so we can log it.